Source/Linux/Utils/ELFContainer.cpp

```diff
@@ -37,40 +64,43 @@ namespace ELFLoader {
 
   FileSize = ELFHeaderSize;
 
-  std::vector<char> RawFile(FileSize);
-  ELFFile.read(RawFile.data(), FileSize);
+  std::vector<char> RawFile(ELFHeaderSize);
+  ELFFile.read(RawFile.data(), ELFHeaderSize);
   ELFFile.close();
 
-  uint8_t *Ident = reinterpret_cast<uint8_t*>(&RawFile.at(0));
+  return CheckELFType(reinterpret_cast<uint8_t*>(&RawFile.at(0)));
+}
 
-  if (Ident[EI_MAG0] != ELFMAG0 ||
-      Ident[EI_MAG1] != ELFMAG1 ||
-      Ident[EI_MAG2] != ELFMAG2 ||
-      Ident[EI_MAG3] != ELFMAG3) {
+ELFContainer::ELFType ELFContainer::GetELFType(int FD) {
+  // We don't know the state of the FD coming in since this might be a guest tracked FD.
+  // Need to be extra careful here not to adjust file offsets and status flags.
+  //
+  // Can't use dup since that makes the FD have the same file description backing both FDs.
```

Comment:
This one is a detail of how Linux implement file handles.
Your typical `FD` is officially a `File Descriptor`.
`FD` is an integer that points to a structure in the kernel called a `File Description`.
The `File Descriptor` is distinct from the `File Description` in that you can have multiple Descriptors pointing to a single Description.

The `File Description` holds all of the state of the file, including locks, read/write position, and other various data.

An example in this case is that you can use `dup` to duplicate the `File Descriptor`, but it still points to the same `File Description`. So if you seek in one FD, the duplicated FD has its position updated.

ex
```cpp
int FD1 = open(...); // Open an FD
lseek(FD1, 512, SEEK_SET); // Offset FD1 to 512
int FD2 = dup(FD1); // Duplicate the file descriptor
lseek(FD2, 0, SEEK_SET); // Offset FD2 to 0

off_t Off1 = lseek(FD1, 0, SEEK_CUR);
off_t Off2 = lseek(FD2, 0, SEEK_CUR);
assert(Off1 == Off2, "This will match");
```

`FD` will always be referring to the integer. `File Description` always referring to the backend data structure.
Which this is all just a detail about why we can't use dup.

Reply:
Is this `File Description` a widely known term? It's the first time I'm hearing about it, and not sure it's a suitable term to use here if it's an obscure term only kernel hackers know about.

Would `Can't use dup since that makes the FD have the same underlying state backing both FDs.` convey the same information?

Reply:
Changed comment.